hands-on-exercise/2-ignite-cli-adv/9-run-prod-docker.md

```diff
@@ -0,0 +1,755 @@
+2. You define a slim new image (the second stage) in which you only copy the compiled file. This is the image you keep for production. It is small.
+
+The disposable image needs to have Rust of at least version 1.56. Fortunately, there are ready-made images. Pick [`rust:1.64.0`](https://hub.docker.com/layers/library/rust/1.64.0/images/sha256-44ba8b8d8a2993694926cc847e1cce27937550c2e9eade4d9887ba90b2a2063f).
+
+In what device will your key be stored? You do not use hardware keys in this setup. So, when building, you use the [`softsign` extension](https://github.com/iqlusioninc/tmkms/blob/c56b496e1bb1187482d7a6fad23c4566329c951e/src/keyring/providers/softsign.rs) with the syntax `--features=softsign`.
+
+What version of the KMS should you compile to? Find the Tendermint version of your Checkers code:
+
+```sh
+$ grep tendermint/tendermint go.mod
+```
+
+It should return something like:
+
+```txt
+github.com/tendermint/tendermint v0.34.19
+```
+
+Because here it is version 0.34, it is a good idea to use the KMS from [version 0.10.0](https://github.com/iqlusioninc/tmkms/blob/main/CHANGELOG.md). At the time of writing, **version 0.12.2** still seems to support Tendermint v0.34. Pick this one.
+
+With these requisites, you can create the staged Docker image:
+
+```Dockerfile [https://github.com/cosmos/b9-checkers-academy-draft/blob/run-prod/Dockerfile-ubuntu-tmkms]
+FROM --platform=linux rust:1.64.0 AS builder
+ARG BUILDARCH
+
+ENV LOCAL=/usr/local
+ENV RUSTFLAGS=-Ctarget-feature=+aes,+ssse3
+ENV TMKMS_VERSION=v0.12.2
+
+RUN apt-get update
+RUN apt-get install libusb-1.0-0-dev --yes
+
+WORKDIR /root
+RUN git clone https://github.com/iqlusioninc/tmkms.git
+WORKDIR /root/tmkms
+RUN git checkout ${TMKMS_VERSION}
+RUN cargo build --release --features=softsign
+
+FROM --platform=linux ubuntu:22.04
+
+COPY --from=builder /root/tmkms/target/release/tmkms ${LOCAL}/bin
+
+ENTRYPOINT [ "tmkms" ]
+```
+
+Note how the production stage is only three lines. Build the image as usual with:
+
+```sh
+$ docker build -f Dockerfile-ubuntu-tmkms . -t tmkms_i
+```
+
+Now when you run it:
+
+```sh
+$ docker run --rm -it tmkms_i
+```
+
+It returns you information about usage. You built Tendermint Key Management System.
+
+## Blockchain elements
+
+Each container needs access to its own private information, such as keys and genesis. To facilitate data access and separation between containers, you create folders that will map as a volume to the default `/root/.checkers`.
+
+```sh
+$ mkdir -p docker/kms-alice
+$ mkdir -p docker/node-carol
+$ mkdir -p docker/sentry-alice
+$ mkdir -p docker/sentry-bob
+$ mkdir -p docker/val-alice
+$ mkdir -p docker/val-bob
+```
+
+For instance, when running a container for `val-alice`, you will create the mapping with a command like:
+
+```sh
+$ docker run ... -v $(pwd)/docker/val-alice:/root/.checkers checkersd_i ...
```

Comment:
This is not super clear.

Reply:
I will add text to say this is a heads-up for later things.